File: src/cli/public-storage-adapter.ts
/**
 * public-storage-adapter.ts
 *
 * Standalone file — copy this to your task-executor project.
 * Zero dependencies on the rest of yaml-flow.
 *
 * Provides:
 *   - KindValueRef      wire format: ::kind::value
 *   - parseRef()        parse a ::kind::value string
 *   - serializeRef()    produce a ::kind::value string
 *   - BlobStorage       read/write interface
 *   - blobStorageForRef resolve a ref to its BlobStorage backend
 *   - ExecutionRef      portable invocation descriptor (inlined, stays standalone)
 *   - TaskCallback      how to report task completion back to the board
 *   - reportComplete()  call from executor on success
 *   - reportFailed()    call from executor on failure
 *
 * Supported storage kinds:
 *   fs-path   — ref.value is an absolute file path; reads/writes via node:fs
 *
 * Supported callback transports (via ExecutionRef.howToRun):
 *   local-node     — invoke board CLI as a child Node process
 *   http:post      — HTTP POST to a board endpoint
 *
 * Usage:
 *   import { parseRef, blobStorageForRef, reportComplete, reportFailed } from './public-storage-adapter.js';
 *
 *   const { source_def, callback } = JSON.parse(blobStorageForRef(inRef).read(inRef.value));
 *   // ... do work, write to outRef ...
 *   reportComplete(callback, outRef);
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { spawnSync } from 'node:child_process';

// ============================================================================
// KindValueRef
// ============================================================================

export interface KindValueRef {
  readonly kind: string;
  readonly value: string;
}

/** Parse a wire-format ref string (::kind::value) into a KindValueRef. */
export function parseRef(s: string): KindValueRef {
  if (!s.startsWith('::')) throw new Error(`Invalid ref format (expected ::kind::value): ${s}`);
  const inner = s.slice(2);
  const idx = inner.indexOf('::');
  if (idx === -1) throw new Error(`Invalid ref format (expected ::kind::value): ${s}`);
  return { kind: inner.slice(0, idx), value: inner.slice(idx + 2) };
}

/** Serialize a KindValueRef to the wire format: ::kind::value */
export function serializeRef(ref: KindValueRef): string {
  return `::${ref.kind}::${ref.value}`;
}

// ============================================================================
// BlobStorage
// ============================================================================

export interface BlobStorage {
  /** Returns content string, or null if not found. */
  read(key: string): string | null;
  /** Write content at key. */
  write(key: string, content: string): void;
  /** Returns true if a blob exists at key. */
  exists(key: string): boolean;
  /** Delete the blob at key. No-op if it does not exist. */
  remove(key: string): void;
}

// ============================================================================
// fs-path backend — key IS the absolute file path
// ============================================================================

function createFsPathBlobStorage(): BlobStorage {
  return {
    read(key: string): string | null {
      if (!fs.existsSync(key)) return null;
      try { return fs.readFileSync(key, 'utf-8'); } catch { return null; }
    },
    write(key: string, content: string): void {
      fs.mkdirSync(path.dirname(key), { recursive: true });
      fs.writeFileSync(key, content, 'utf-8');
    },
    exists(key: string): boolean {
      return fs.existsSync(key);
    },
    remove(key: string): void {
      try { if (fs.existsSync(key)) fs.unlinkSync(key); } catch { /* best-effort */ }
    },
  };
}

// ============================================================================
// blobStorageForRef
// ============================================================================

/**
 * Resolve a KindValueRef to its BlobStorage backend.
 * Throws a clear error for unrecognised kinds.
 */
export function blobStorageForRef(ref: KindValueRef): BlobStorage {
  switch (ref.kind) {
    case 'fs-path': return createFsPathBlobStorage();
    default: throw new Error(`Unsupported storage kind: "${ref.kind}". Supported kinds: fs-path`);
  }
}

// ============================================================================
// TaskCallback — how a task-executor reports results back to the board
// ============================================================================

/**
 * Portable invocation descriptor for the board CLI back-channel.
 * Inlined here so this file stays standalone (zero deps on yaml-flow internals).
 * Shape matches ExecutionRef in execution-interface.ts — keep in sync.
 *
 * Supported howToRun values for TaskCallback.via:
 *   local-node   — invoke board CLI as: node [tsx?] <whatToRun.value> <cmd> [...argv]
 *   http:post    — POST to <whatToRun.value> with a JSON body
 */
export interface ExecutionRef {
  /** Optional human-readable label. Not used for dispatch. */
  meta?: string;
  /** Transport / runtime kind. */
  howToRun: 'local-node' | 'local-python' | 'local-process' | 'http:post' | 'http:get' | 'built-in';
  /** Address of the target in ::kind::value wire form (e.g. ::fs-path::/path/to/cli.js). */
  whatToRun: string;
  /** Opaque executor config stored with the ref. */
  extra?: Record<string, unknown>;
}

/**
 * Describes how the board wants to receive task completion callbacks.
 * Baked into the inRef payload as { source_def, callback }.
 * The executor treats `token` as opaque and passes it back unchanged.
 */
export interface TaskCallback {
  /** Opaque routing token — generated by the board, passed back unchanged. */
  token: string;
  /** Delivery mechanism — an ExecutionRef pointing at the board CLI or endpoint. */
  via: ExecutionRef;
}

/**
 * Extract the path/url value from a whatToRun ::kind::value wire string.
 * Falls back to the raw string if it isn’t in ::kind::value form.
 */
function _parseWhatToRun(whatToRun: string): string {
  try { return parseRef(whatToRun).value; } catch { return whatToRun; }
}

/**
 * Resolve the Node invocation for a local board CLI script.
 * If the path ends in .ts (dev mode), attempts to locate tsx alongside it;
 * otherwise assumes it’s a compiled .js and invokes directly with node.
 */
function _resolveLocalNodeInvocation(scriptPath: string): { cmd: string; args: string[] } {
  if (!scriptPath.endsWith('.ts')) {
    return { cmd: process.execPath, args: [scriptPath] };
  }
  // Dev path: look for tsx next to node_modules relative to the script
  const candidates = [
    path.join(path.dirname(scriptPath), '..', '..', 'node_modules', 'tsx', 'dist', 'cli.mjs'),
    path.join(path.dirname(scriptPath), '..', '..', 'node_modules', '.bin', 'tsx'),
  ];
  const tsx = candidates.find(p => fs.existsSync(p));
  if (tsx) return { cmd: process.execPath, args: [tsx, scriptPath] };
  return { cmd: 'npx', args: ['tsx', scriptPath] };
}

/**
 * Report successful task completion back to the board.
 * Call this from a task-executor after writing the result to outRef.
 */
export function reportComplete(callback: TaskCallback, outRef: KindValueRef): void {
  const { token, via } = callback;
  if (via.howToRun === 'local-node' || via.howToRun === 'local-process') {
    const scriptPath = _parseWhatToRun(via.whatToRun);
    const { cmd, args } = _resolveLocalNodeInvocation(scriptPath);
    const result = spawnSync(cmd, [...args, 'source-data-fetched',
      '--ref', serializeRef(outRef), '--token', token,
    ], { encoding: 'utf-8', windowsHide: true });
    if (result.status !== 0) {
      throw new Error(`reportComplete: board CLI exited ${result.status}: ${result.stderr?.trim()}`);
    }
    return;
  }
  if (via.howToRun === 'http:post') {
    const url = _parseWhatToRun(via.whatToRun);
    const body = JSON.stringify({ status: 'complete', ref: serializeRef(outRef), token });
    _httpPostSync(url, body);
    return;
  }
  throw new Error(`reportComplete: unsupported via.howToRun "${via.howToRun}"`);
}

/**
 * Report task failure back to the board.
 * Call this from a task-executor instead of writing to outRef.
 */
export function reportFailed(callback: TaskCallback, reason: string): void {
  const { token, via } = callback;
  if (via.howToRun === 'local-node' || via.howToRun === 'local-process') {
    const scriptPath = _parseWhatToRun(via.whatToRun);
    const { cmd, args } = _resolveLocalNodeInvocation(scriptPath);
    const result = spawnSync(cmd, [...args, 'source-data-fetch-failure',
      '--token', token, '--reason', reason,
    ], { encoding: 'utf-8', windowsHide: true });
    if (result.status !== 0) {
      throw new Error(`reportFailed: board CLI exited ${result.status}: ${result.stderr?.trim()}`);
    }
    return;
  }
  if (via.howToRun === 'http:post') {
    const url = _parseWhatToRun(via.whatToRun);
    const body = JSON.stringify({ status: 'failed', reason, token });
    _httpPostSync(url, body);
    return;
  }
  throw new Error(`reportFailed: unsupported via.howToRun "${via.howToRun}"`);
}

/** Synchronous HTTP POST using a child node process (keeps this file free of async). */
function _httpPostSync(url: string, body: string): void {
  const script = `
    const {request} = require(new URL('${url}').protocol === 'https:' ? 'https' : 'http');
    const h = ${JSON.stringify({ 'Content-Type': 'application/json', 'Content-Length': Buffer.byteLength(body) })};
    const u = new URL('${url}');
    const req = request({hostname:u.hostname,port:u.port,path:u.pathname+u.search,method:'POST',headers:h});
    req.on('error', e => { process.stderr.write(e.message); process.exit(1); });
    req.write(${JSON.stringify(body)});
    req.end();
  `;
  const result = spawnSync(process.execPath, ['-e', script], { encoding: 'utf-8', windowsHide: true });
  if (result.status !== 0) throw new Error(`http-post failed: ${result.stderr?.trim()}`);
}

File: pycli/sub/public_storage_adapter.py
"""
public_storage_adapter.py

Standalone file — copy this to your task-executor project.
Zero dependencies on the rest of yaml-flow internals.

Provides:
  - KindValueRef        wire format: ::kind::value
  - parse_ref()         parse a ::kind::value string
  - serialize_ref()     produce a ::kind::value string
  - BlobStorage         read/write interface (protocol class)
  - blob_storage_for_ref()  resolve a ref to its BlobStorage backend
  - TaskCallback        how to report task completion back to the board
  - report_complete()   call from executor on success
  - report_failed()     call from executor on failure

Supported storage kinds:
  fs-path   — ref.value is an absolute file path; reads/writes via os/pathlib

Supported callback transports (via ExecutionRef.howToRun):
  local-node     — invoke board CLI as a child Node process
  local-python   — invoke board pycli as a child Python process
  http:post      — HTTP POST to a board endpoint

Usage:
  from pycli.sub.public_storage_adapter import (
      parse_ref, serialize_ref, blob_storage_for_ref,
      report_complete, report_failed,
  )

  in_ref = parse_ref(in_ref_str)
  storage = blob_storage_for_ref(in_ref)
  envelope = json.loads(storage.read(in_ref.value))
  callback = envelope.get("callback")
  # ... do work, write to out_ref ...
  report_complete(callback, out_ref)
"""
from __future__ import annotations

import json
import os
import subprocess
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Optional, Protocol
from urllib.request import Request, urlopen


# ============================================================================
# KindValueRef
# ============================================================================

@dataclass(frozen=True)
class KindValueRef:
    kind: str
    value: str


def parse_ref(s: str) -> KindValueRef:
    """Parse a wire-format ref string (::kind::value) into a KindValueRef."""
    if not s.startswith("::"):
        raise ValueError(f"Invalid ref format (expected ::kind::value): {s}")
    inner = s[2:]
    idx = inner.find("::")
    if idx < 0:
        raise ValueError(f"Invalid ref format (expected ::kind::value): {s}")
    return KindValueRef(kind=inner[:idx], value=inner[idx + 2:])


def serialize_ref(ref: KindValueRef) -> str:
    """Serialize a KindValueRef to the wire format: ::kind::value"""
    return f"::{ref.kind}::{ref.value}"


# ============================================================================
# BlobStorage
# ============================================================================

class BlobStorage(Protocol):
    def read(self, key: str) -> Optional[str]: ...
    def write(self, key: str, content: str) -> None: ...


class _FsPathBlobStorage:
    """fs-path backend — key IS the absolute file path."""

    def read(self, key: str) -> Optional[str]:
        if not os.path.exists(key):
            return None
        return Path(key).read_text(encoding="utf-8")

    def write(self, key: str, content: str) -> None:
        os.makedirs(os.path.dirname(key), exist_ok=True)
        Path(key).write_text(content, encoding="utf-8")


def blob_storage_for_ref(ref: KindValueRef) -> _FsPathBlobStorage:
    """Resolve a KindValueRef to its BlobStorage backend."""
    if ref.kind == "fs-path":
        return _FsPathBlobStorage()
    raise ValueError(f'Unsupported storage kind: "{ref.kind}". Supported kinds: fs-path')


# ============================================================================
# TaskCallback — how a task-executor reports results back to the board
# ============================================================================

@dataclass
class ExecutionRef:
    """Portable invocation descriptor (mirrors execution-interface.ts)."""
    howToRun: str
    whatToRun: str
    meta: Optional[str] = None
    extra: Optional[dict[str, Any]] = None


@dataclass
class TaskCallback:
    """Describes how the board wants to receive task completion callbacks."""
    token: str
    via: ExecutionRef


def _parse_task_callback(raw: dict[str, Any]) -> TaskCallback:
    """Parse a raw dict (from JSON envelope) into a TaskCallback."""
    via_raw = raw.get("via") or {}
    return TaskCallback(
        token=str(raw.get("token", "")),
        via=ExecutionRef(
            howToRun=str(via_raw.get("howToRun", "")),
            whatToRun=str(via_raw.get("whatToRun", "")),
            meta=via_raw.get("meta"),
            extra=via_raw.get("extra"),
        ),
    )


def _parse_what_to_run(what_to_run: str) -> str:
    """Extract the path/url value from a whatToRun ::kind::value wire string."""
    try:
        return parse_ref(what_to_run).value
    except Exception:
        return what_to_run


def _notify_channel_from_via(via: ExecutionRef) -> Optional[str]:
    if via.extra and isinstance(via.extra.get("notifyChannel"), str):
        ch = via.extra["notifyChannel"]
        return ch if ch else None
    return None


def _run_subprocess_hidden(cmd: list[str]) -> subprocess.CompletedProcess[str]:
    """Run a subprocess without popping a cmd window on Windows."""
    kwargs: dict[str, Any] = {
        "capture_output": True,
        "text": True,
        "shell": False,
    }
    if os.name == "nt":
        startupinfo = subprocess.STARTUPINFO()
        startupinfo.dwFlags |= subprocess.STARTF_USESHOWWINDOW
        kwargs["startupinfo"] = startupinfo
        kwargs["creationflags"] = subprocess.CREATE_NO_WINDOW
    return subprocess.run(cmd, **kwargs)


def report_complete(callback: Any, out_ref: KindValueRef) -> None:
    """
    Report successful task completion back to the board.
    Call this from a task-executor after writing the result to out_ref.

    Accepts either a TaskCallback or a raw dict (from JSON envelope).
    """
    cb = callback if isinstance(callback, TaskCallback) else _parse_task_callback(callback)
    token = cb.token
    via = cb.via

    if via.howToRun in ("local-node", "local-process"):
        script_path = _parse_what_to_run(via.whatToRun)
        notify_channel = _notify_channel_from_via(via)
        cmd = ["node", script_path, "source-data-fetched",
               "--ref", serialize_ref(out_ref),
               "--token", token]
        if notify_channel:
            cmd.extend(["--notify-channel", notify_channel])
        result = _run_subprocess_hidden(cmd)
        if result.returncode != 0:
            msg = (result.stderr or result.stdout or "callback failed").strip()
            raise RuntimeError(f"report_complete: board CLI exited {result.returncode}: {msg}")
        return

    if via.howToRun == "local-python":
        script_path = _parse_what_to_run(via.whatToRun)
        notify_channel = _notify_channel_from_via(via)
        # Use unprefixed command — board pycli decodes base_ref from token automatically.
        cmd = [sys.executable, script_path, "source-data-fetched",
               "--ref", serialize_ref(out_ref),
               "--token", token]
        if notify_channel:
            cmd.extend(["--notify-channel", notify_channel])
        result = _run_subprocess_hidden(cmd)
        if result.returncode != 0:
            msg = (result.stderr or result.stdout or "callback failed").strip()
            raise RuntimeError(f"report_complete: board pycli exited {result.returncode}: {msg}")
        return

    if via.howToRun == "http:post":
        url = _parse_what_to_run(via.whatToRun)
        body = json.dumps({"status": "complete", "ref": serialize_ref(out_ref), "token": token}).encode("utf-8")
        req = Request(url, method="POST", data=body, headers={"Content-Type": "application/json"})
        with urlopen(req, timeout=30):
            return

    raise ValueError(f'report_complete: unsupported via.howToRun "{via.howToRun}"')


def report_failed(callback: Any, reason: str) -> None:
    """
    Report task failure back to the board.
    Call this from a task-executor instead of writing to out_ref.

    Accepts either a TaskCallback or a raw dict (from JSON envelope).
    """
    cb = callback if isinstance(callback, TaskCallback) else _parse_task_callback(callback)
    token = cb.token
    via = cb.via

    if via.howToRun in ("local-node", "local-process"):
        script_path = _parse_what_to_run(via.whatToRun)
        notify_channel = _notify_channel_from_via(via)
        cmd = ["node", script_path, "source-data-fetch-failure",
               "--token", token,
               "--reason", reason]
        if notify_channel:
            cmd.extend(["--notify-channel", notify_channel])
        result = _run_subprocess_hidden(cmd)
        if result.returncode != 0:
            msg = (result.stderr or result.stdout or "callback failed").strip()
            raise RuntimeError(f"report_failed: board CLI exited {result.returncode}: {msg}")
        return

    if via.howToRun == "local-python":
        script_path = _parse_what_to_run(via.whatToRun)
        notify_channel = _notify_channel_from_via(via)
        # Use unprefixed command — board pycli decodes base_ref from token automatically.
        cmd = [sys.executable, script_path, "source-data-fetch-failure",
               "--token", token,
               "--reason", reason]
        if notify_channel:
            cmd.extend(["--notify-channel", notify_channel])
        result = _run_subprocess_hidden(cmd)
        if result.returncode != 0:
            msg = (result.stderr or result.stdout or "callback failed").strip()
            raise RuntimeError(f"report_failed: board pycli exited {result.returncode}: {msg}")
        return

    if via.howToRun == "http:post":
        url = _parse_what_to_run(via.whatToRun)
        body = json.dumps({"status": "failed", "reason": reason, "token": token}).encode("utf-8")
        req = Request(url, method="POST", data=body, headers={"Content-Type": "application/json"})
        with urlopen(req, timeout=30):
            return

    raise ValueError(f'report_failed: unsupported via.howToRun "{via.howToRun}"')
